File: 2815-max-pair-sum-in-an-array/2815-max-pair-sum-in-an-array.py
class Solution:
    def maxSum(self, nums: List[int]) -> int:
        max_by_digit = defaultdict(int)
        max_sum = -1

        for num in nums:
            digit = max(str(num))

            if digit in max_by_digit:
                max_sum = max(max_sum, max_by_digit[digit] + num)

            max_by_digit[digit] = max(max_by_digit[digit], num)

        return max_sum

File: 2815-max-pair-sum-in-an-array/2815-max-pair-sum-in-an-array.cpp
class Solution {
public:
    int maxSum(vector<int>& nums) {
        int res = 0;
        for(int i = 0; i<nums.size(); i++){
            string first = to_string(nums[i]);
            sort(begin(first),end(first));
            
            for(int j = i+1; j<nums.size(); j++){
                string sec = to_string(nums[j]);
                sort(begin(sec),end(sec));
                if(first[first.size()-1] == sec[sec.size()-1]){
                    res = max(res,nums[i]+nums[j]);
                }
            }
        }
        
        return res == 0 ? -1 : res;
    }
};